0010-keccak-syscall.md

```diff
@@ -0,0 +1,402 @@
+used while fetching the memory.
+
+Note that as the syscall instruction must be encoded in a circuit and its
+behavior depends on the value in the register `v0`, we must represent its effect
+using a polynomial constraining the value `v0`. We can easily do that using a
+vanishing polynomial on the values different than the one expected.
+For instance, let's suppose `v0` is `sysFnctl` and `sysFnctl` behavior is represented
+by the polynomial `P(v0, a0, a1, a3)`, we will use the polynomial:
+```
+P_syscall_fnctl(v0, a0, a1, a3) = \
+  (v0 - 4090) *
+  (v0 - 4045) *
+  (v0 - 4120) *
+  (v0 - 4246) *
+  (v0 - 4003) *
+  (v0 - 4004) *
+  P(v0, a0, a1, a3)
+```
+
+to activate the `P_syscall_fnctl` behavior when `v0 = 4055` and deactivate it
+when it has a different value. Reduction to polynomials of at most degree `2` can be
+done using [quadracization](./0008-kimchi-polynomial-quadracization.md).
+
+We suppose quadracization is going to be used, and therefore, we will not talk
+about the degree of the different polynomial constraints, and the detailled
+description will stay high level using the API described above, with some
+additional tweaks if necessary for a specific syscall behavior.
+
+Regarding the pre-oracle data, it will be loaded as part of the witness given to
+the prover.
+In the context of generating a proof using the zkVM, the program will first be
+run outside of the zkVM, and therefore we know which data and how much data will
+be requested by the program.
+A new vector of bytes can be added to the Witness structure, mimicking the work
+done for the memory, which will keep track of the data requested by the program.
+The [CLI](./cli.md) integration will take care of providing and populating
+correctly the Witness instance.
+
+Regarding the encoding, we must represent the preimage-key as
+a field element of the scalar field of the pairing-friendly curve BN128, which
+is 254 bits. The keys are 32 bytes long as it is the result of a KECCAK execution and
+therefore is too large to be saved in a field element.
+According to the [CLI RFC](./cli.md), the first byte of the hash must be dropped to be
+replaced by the request type, leaving therefore 248 bits to be encoded in a
+field element, which fits into an element of the scalar field of BN128.
+The first byte can be used to encode the table ID the pre-image data are going
+to be in.
+For the rest of the document, it will be supposed that only type 2 pre-image
+keys must be supported, i.e. the table ID will be considered to be a certain
+arbitrary constant.
+
+Regarding the the pre-image data, it is important to note that it can be of any
+size. The efficiency of a representation and encoding in a lookup table will
+depend on the sizes of the data and also its number, and therefore it is
+left to the kind of applications the VM is running.
+However, there are some optimisations that can be used.
+As the key is only 31 bytes, i.e. 248 bits, there are still 5 bits of data that
```

Comment:
I don't follow where the 5 bits come from. Is that using the space left in the byte encoding the key type which we assume to be 0x02?

Reply:
Yes

Reply:
Would you mind making that clearer?